### PR TITLE
[Issue #39] CreateConversations

### PR DESCRIPTION
**Pull request by russnicoletti**
This is the initial commit for implementing the CreateConversations endpoint (part of the Data Array extension). This commit is focused on validating the input.

My objective of creating this PR is to get feedback on my approach.

### CODE REVIEW

File: src/extensions/dataarray.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      }
+
+      let properties = obj.components.toString();
+      requiredProperties.forEach(property => {
+        if (properties.indexOf(property) == -1) {
+          if (missingProperties.indexOf(property) === -1) {
+            missingProperties.push(property);
+          }
+        }
+      });
```
Comment (ferjm) [inline src/extensions/dataarray.js]: An alternative approach.

```javascript 
requiredProperties.forEach(property => {
  if (!obj.components[property]) {
    missingProperties[property] = true;
  }
})
```

File: src/extensions/dataarray.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    // Required properties in each object:
+    //   - Datastream
+    //   - phenomenonTime
+    //   - result
+    //   - featureOfInterest
+    let requiredProperties = ['phenomenonTime', 'result', 'FeatureOfInterest'];
+    var validateObject = function(obj) {
+      if (obj.Datastream === undefined) {
+        if (missingProperties.indexOf(datastream) === -1) {
```
Comment (ferjm) [inline src/extensions/dataarray.js]: And in any case, I wouldn't use an Array for `missingProperties`. I would use an Object, set the property whenever is missing with a dummy value like `true` and do a `Object.keys(missingProperties)` to get the array of missing properties. This way the conditions are simpler:

```javascript
  if (!obj[datastream]) {
    missingProperties[datastream] = true;
  }
```

File: src/extensions/dataarray.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    // Validate input.
+    // 1) The body of request must be an array.
+    if (Object.prototype.toString.call(body) !== '[object Array]') {
+      ERR.ApiError(res, 400, ERR.ERRNO_BAD_REQUEST,
+                   ERR.BAD_REQUEST, 'Input must be an array');
+    }
+
+    // 2) Valildate that each object in the input has the required properties.
+    let missingProperties = [];
```
Comment (ferjm) [inline src/extensions/dataarray.js]: To be honest, I wouldn't worry too much about checking the entire request looking for all the missing properties. I would simply bail out as soon as I find the first missing property (i.e. throwing an exception inside `validateObject`).

Anyway, if you want to check all the the missing properties on the entire request, try to do it in a more functional way, please. For example, you could explicitly pass `missingProperties` to `validateObject` or make `validateObject` return an Array so you can do `missingProperties.push(validateObject(obj))`. Leaving `missingProperties` in the global scope of the module seems a bit dangerous to me.

File: src/extensions/dataarray.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    body.forEach(obj => {
+      console.log('validating object...');
+      dump(obj);
+      validateObject(obj);
+    });
+
```
Comment (ferjm) [inline src/extensions/dataarray.js]: I think we also need to check for the `dataArray` property

File: src/extensions/dataarray.js (hunk 3 above)
Comment (russnicoletti) [inline src/extensions/dataarray.js]: I also wasn't sure if it was worth the effort to check the entire request. And now I've decided it is not worth the effort so the code now bails when it detects the first missing property. This makes the code more straightforward and efficient.

File: src/extensions/data_array.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    // 2) Validate that each object in the input has the required properties.
+    for (let i = 0; i < body.length; i++) {
+      try {
+        validateBody(body[i]);
+      } catch(ex) {
+        ERR.ApiError(res, ERR.errnos[ex.errno], ex.errno, ex.name, ex.errors);
+        return;
+      }
+    }
+
```
Comment (ferjm) [inline src/extensions/data_array.js]: I would also check that `components` and `dataArray` have the same length.

File: test/test_create_observations.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+const endpoint = '/v1.0/' + createObservations;
+const server   = supertest.agent(app.listen(8889));
+
+const postSuccess = (body, done) => {
+  server.post(endpoint).send(body)
+  .expect(201)
+  .end((err, res) => {
+    should.not.exist(err);
+    res.status.should.be.equal(201);
```
Comment (albertopq) [inline test/test_create_observations.js]: Can we check here that in the body we have a list of the Observations urls that we are expecting?

File: test/constants.js [hunk 7]
```diff
@@ -105,6 +105,56 @@ let testEntities = {
   }
 };
 
+let testCreateObservationsRequestBodies = {
+  CreateObservationsRequestOneObservation:
+  [
+    {
+      Datastream: { '@iot.id': 1 },
```
Comment (albertopq) [inline test/constants.js]: Are we assuming that a Datastream with id: 1 exists?

File: test/test_create_observations.js (hunk 6 above)
Comment (russnicoletti) [inline test/test_create_observations.js]: Yes, although I haven't written that code yet. Currently, the CreateObserverations implementation only validates the input; it doesn't create the observation entities.

File: test/constants.js (hunk 7 above)
Comment (russnicoletti) [inline test/constants.js]: I'm not sure. I believe it should exist. So I guess the '1' is a place holder for the actual iot.id of the Datastream that the test will create before testing the CreateObjerverations endpoint.

File: test/test_create_observations.js (hunk 6 above)
Comment (albertopq) [inline test/test_create_observations.js]: Yeah, that is true. Lets just land this then, as we are only validating the request, and that is being tested. Thanks Russ!

**Pull request merged**